docs/dev/table/functions/udfs.md

```diff
@@ -22,322 +22,768 @@ specific language governing permissions and limitations
+  "LEFT JOIN LATERAL TABLE(SplitFunction(myField)) ON TRUE");
+
+// rename fields of the function in SQL
+env.sqlQuery(
+  "SELECT myField, newWord, newLength " +
+  "FROM MyTable " +
+  "LEFT JOIN LATERAL TABLE(SplitFunction(myField)) AS T(newWord, newLength) ON TRUE");
 
 {% endhighlight %}
+</div>
 
-There are many ways to define a Python table function besides extending the base class `TableFunction`.
-Please refer to the [Python Table Function]({{ site.baseurl }}/dev/table/python/python_udfs.html#table-functions) documentation for more details.
- 
+<div data-lang="Scala" markdown="1">
+{% highlight scala %}
+import org.apache.flink.table.annotation.DataTypeHint
+import org.apache.flink.table.annotation.FunctionHint
+import org.apache.flink.table.api._
+import org.apache.flink.table.functions.TableFunction
+import org.apache.flink.types.Row
+
+@FunctionHint(output = @DataTypeHint("ROW<word STRING, length INT>"))
+class SplitFunction extends TableFunction[Row] {
+
+  def eval(str: String): Unit = {
+    // use collect(...) to emit a row
+    str.split(" ").foreach(s => collect(Row.of(s, s.length)))
+  }
+}
+
+val env = TableEnvironment.create(...)
+
+// call function "inline" without registration in Table API
+env
+  .from("MyTable")
+  .joinLateral(call(classOf[SplitFunction], $"myField")
+  .select($"myField", $"word", $"length")
+env
+  .from("MyTable")
+  .leftOuterJoinLateral(call(classOf[SplitFunction], $"myField"))
+  .select($"myField", $"word", $"length")
+
+// rename fields of the function in Table API
+env
+  .from("MyTable")
+  .leftOuterJoinLateral(call(classOf[SplitFunction], $"myField").as("newWord", "newLength"))
+  .select($"myField", $"newWord", $"newLength")
+
+// register function
+env.createTemporarySystemFunction("SplitFunction", classOf[SplitFunction])
+
+// call registered function in Table API
+env
+  .from("MyTable")
+  .joinLateral(call("SplitFunction", $"myField"))
+  .select($"myField", $"word", $"length")
+env
+  .from("MyTable")
+  .leftOuterJoinLateral(call("SplitFunction", $"myField"))
+  .select($"myField", $"word", $"length")
+
+// call registered function in SQL
+env.sqlQuery(
+  "SELECT myField, word, length " +
+  "FROM MyTable, LATERAL TABLE(SplitFunction(myField))");
+env.sqlQuery(
+  "SELECT myField, word, length " +
+  "FROM MyTable " +
+  "LEFT JOIN LATERAL TABLE(SplitFunction(myField)) ON TRUE")
+
+// rename fields of the function in SQL
+env.sqlQuery(
+  "SELECT myField, newWord, newLength " +
+  "FROM MyTable " +
+  "LEFT JOIN LATERAL TABLE(SplitFunction(myField)) AS T(newWord, newLength) ON TRUE")
+
+{% endhighlight %}
 </div>
+
 </div>
 
-{% top %}
+If you intend to implement functions in Scala, do not implement a table function as a Scala `object`. Scala `object`s are singletons and will cause concurrency issues.
 
+If you intend to implement or call functions in Python, please refer to the [Python Table Functions](../python/python_udfs.html#table-functions) documentation for more details.
```

Comment:
Out of scope for this change, but I don't like how the python docs are seperated from everything else.

Reply:
I also don't like that but the JVM APIs and Python differ a lot. Python has own annotations, a subset of the data types (that even still go through type information internally), and just basic function support. So not much in common with what is documented here. In the previous version, the documentation was even duplicated which is also hard to maintain.